### PR TITLE
regression of Creator check, #20537

### PR DESCRIPTION
**Pull request by patriknw**
I have not been able to reproduce it yet, but this might be a starting point for writing a reproducer. @guidomedina can you help me to modify this test so that it matches your way of defining the Creator, and fails the test?

### CODE REVIEW

**Comment by ktoso** [comment]
Used Java versions may also be interesting for this hm...

**Comment by guidomedina** [comment]
I'm using Java 8u92 

I suspect this wouldn't fail (creating with creator inside a method):

```
context().actorOf(Props.create(priceProcessorCreator).withMailbox("bounded-mailbox-" + 2048 / CONCURRENCY_LEVEL));
```

But this will fail (creating with creator inside a lambda):
```
IntStream.range(0, CONCURRENCY_LEVEL).
      mapToObj(i -> context().actorOf(Props.create(priceProcessorCreator).withMailbox("bounded-mailbox-" + 2048 / CONCURRENCY_LEVEL))).
      collect(toCollection(() -> new ObjectArrayList<>(CONCURRENCY_LEVEL)))
```

**Comment by guidomedina** [comment]
Let me write a simple reproducer, just so that you can have it at hand as starting point.

**Comment by patriknw** [comment]
thanks

**Comment by guidomedina** [comment]
I wasn't able to reproduce it for simple and straight forward code but was able to fix it for my code by making the protected constructor of my creator either public or package.

I'm not sure if this helps but I can say in my case that I was doing "something wrong" so IMHO I think there is no issue with Akka but my old code.

**Comment by patriknw** [comment]
ok, so it doesn't fail just by placing stuff into a lambda?

I'm fine with that resolution, but leave it open for a few days of thinking time.

**Comment by guidomedina** [comment]
Reproduced with the following, even though it compiles in java, Akka class loader struggles with it, but in my case I'm getting a different exception but I think I can conclude that it is an access problem, I think should be documented, nothing to fix IMHO:

```
import akka.actor.*;
import akka.japi.Creator;

import java.util.ArrayList;
import java.util.stream.IntStream;

import static java.util.stream.Collectors.toCollection;

public class Reproducer extends UntypedActor {

  static final class ReproducerCreator implements Creator<Reproducer> {

    final boolean create;

    private ReproducerCreator(boolean create) {
      this.create = create;
    }

    @Override
    public Reproducer create() throws Exception {
      return new Reproducer(create);
    }
  }

  public Reproducer(boolean create) {
    if (create) {
      final ReproducerCreator creator = new ReproducerCreator(false);
      IntStream.range(0, 4).mapToObj(i -> context().actorOf(Props.create(creator))).collect(toCollection(ArrayList::new));
      System.out.println("Created multiple actors with creator inside lambda");
    }
  }

  @Override
  public void onReceive(Object message) throws Exception {
  }

  public static void main(String[] args) {
    final ActorSystem system = ActorSystem.create();
    final ReproducerCreator creator = new ReproducerCreator(true);
    system.actorOf(Props.create(creator));
    System.out.println("Created single actor with creator");
  }
}
```

**Comment by guidomedina** [comment]
To make it pass change creator constructor from `private` to something else and it works.

**Comment by patriknw** [comment]
Refs https://github.com/akka/akka/issues/20537

File: akka-actor/src/main/scala/akka/actor/AbstractProps.scala [hunk 1]
```diff
@@ -87,7 +87,7 @@ private[akka] trait AbstractProps {
         if (i == declaredConstructors.length) false
         else {
           val c = declaredConstructors(i)
-          if (c.getParameterCount >= 1 && c.getParameterTypes()(i) == enclosingClass)
+          if (c.getParameterCount >= 1 && c.getParameterTypes()(0) == enclosingClass)
```
Comment (patriknw) [inline akka-actor/src/main/scala/akka/actor/AbstractProps.scala]: @guidomedina Thanks for the reproducer. It was actually a bug.

File: akka-actor-tests/src/test/java/akka/actor/ActorCreationTest.java [hunk 2]
```diff
@@ -287,5 +315,18 @@ public void testPropsUsingCreatorWithoutClass() {
...
   }
 
+  @Test
+  public void testIssue20537Reproducer() {
+    final Issue20537Reproducer.ReproducerCreator creator = new Issue20537Reproducer.ReproducerCreator(false);
+    final Props p = Props.create(creator);
+    assertEquals(Issue20537Reproducer.class, p.actorClass());
+
+    ArrayList<Props> pList = IntStream.range(0, 4).mapToObj(i -> Props.create(creator))
+        .collect(toCollection(ArrayList::new));
```
Comment (patriknw) [inline akka-actor-tests/src/test/java/akka/actor/ActorCreationTest.java]: the lambda doesn't matter for reproducing the issue, but I included that test anyway

File: akka-actor/src/main/scala/akka/actor/AbstractProps.scala (hunk 1 above)
Comment (guidomedina) [inline akka-actor/src/main/scala/akka/actor/AbstractProps.scala]: @patriknw glad I was able to help.

Comment (ktoso) [inline akka-actor/src/main/scala/akka/actor/AbstractProps.scala]: Oh sneaky! Thanks for locating this guys

File: akka-actor-tests/src/test/java/akka/actor/ActorCreationTest.java (hunk 2 above)
Comment (guidomedina) [inline akka-actor-tests/src/test/java/akka/actor/ActorCreationTest.java]: yeah, I wanted to provide a reproducer as close as possible to the original code.

**Comment by patriknw** [comment]
The test failure was unrelated, created issue https://github.com/akka/akka/issues/20601

**Comment by patriknw** [comment]
PLS BUILD

**Comment by ktoso** [comment]
Failure was unrelated timing issue: https://github.com/akka/akka/issues/20607

**Comment by ktoso** [comment]
LGTM

**Pull request merged**